libc/sem.cc

```diff
@@ -76,3 +105,72 @@ int sem_trywait(sem_t* s)
         return libc_error(EAGAIN);
     return 0;
 }
+
+static std::unordered_map<std::string, indirect_semaphore> named_semaphores;
+static mutex named_semaphores_mutex;
+
+OSV_LIBC_API
+sem_t *sem_open(const char *name, int oflag, mode_t mode, unsigned int value)
+{
+    SCOPE_LOCK(named_semaphores_mutex);
+    auto iter = named_semaphores.find(std::string(name));
+    
+    if (iter != named_semaphores.end()) {
+        //opening already named semaphore
+        if (oflag & (O_EXCL | O_CREAT)) {
```

Comment:
Are you sure about O_CREAT here? The Linux manual page suggests that "if  O_CREAT  is specified in oflag, then the semaphore is created if it does not already exist." so it shouldn't be an error to use O_CREAT. Moreover the documentation says "If  both O_CREAT and O_EXCL are specified in oflag, then an error is returned if a semaphore with the  given  name  already exists."

So maybe the right if() should be

```c++
if (oflag & O_EXCL && oflag & O_CREATE)
```
or alternatively
```c++
if (oflag & (O_EXCL | O_CREATE)) == (O_EXCL | O_CREAT))
```
Do you have a test for this?

Reply:
This was more insidious than I thought. I believe your first check would be the correct method. The second could lead to incorrect behavior if the O_CREAT and O_EXCL flags are both set while any other bit is also set. Because of the exact equality check, any other bit set would mean that we would not be testing just for the failure case of O_CREAT and O_EXCL. I adopted the first check and added tests for this check, as well as some new tests for the following portion where new semaphores are created.

Reply:
Thanks. Please push the new version when you're ready.